README.md
# OpenCensus Zenoss Exporter for Node.js

A [Node.js] library intended to be used in [Node.js] applications instrumented
with [OpenCensus] to export stats to [Zenoss].

[Node.js]: https://nodejs.org/ 
[OpenCensus]: https://opencensus.io/
[Zenoss]: https://zenoss.com/

## Status

This library is in an alpha stage, and the API is subject to change.

## Usage

To use Zenoss as your exporter, you will first need a Zenoss API key. Once you
have an API key, add the following to your Node.js application.

For Javascript:
```javascript
const { globalStats } = require("@opencensuscore")
const { ZenossStatsExporter } = require("@zenoss/opencensus-node-exporter")

const exporter = new ZenossStatsExporter({ apiKey: "YOUR-ZENOSS-API-KEY" })
globalStats.registerExporter(exporter)
```

For TypeScript:
```typescript
import { globalStats } from "@opencensus/core";
import { ZenossStatsExporter } from "@zenoss/opencensus-node-exporter"

const exporter = new ZenossStatsExporter({ apiKey: "YOUR-ZENOSS-API-KEY" })
globalStats.registerExporter(exporter)
```

Be sure to install the exporter. For `npm` use the following:
```shell script
npm install @zenoss/opencensus-node-exporter
```

## Options

The following options are available when creating a `ZenossStatsExporter`.

- `address`: Zenoss API address. Default is https://api.zenoss.io.
- `apiKey`: Zenoss API key.
- `source`: Added as a tag to all sent metrics. Recommended.
- `extraTags`: Map of additional tags to add to all sent metrics. Default is {}.
- `period`: How often (in milliseconds) to send metrics to Zenoss. Default is 60000.
- `logger`: Optional logger to be used by the exporter.

## Example Application

The following is a complete example of an application that will write the value
`10` to a measure named `example.com/ten` once per second. Then once per minute
the last value of this metric will be sent to Zenoss with a metric name of
`example.com/ten_last`.

```javascript
const { globalStats, AggregationType, MeasureUnit, TagMap } = require('@opencensus/core')
const { ZenossStatsExporter } = require('@zenoss/opencensus-node-exporter')

// Setup the Zenoss exporter.
const exporter = new ZenossStatsExporter({ apiKey: "YOUR-ZENOSS-API-KEY" })
globalStats.registerExporter(exporter)

// Setup our tags.
const tagKeyApp = { name: "app" }
const tagMapTen = new TagMap()
tagMapTen.set(tagKeyApp, { value: "nodejs-example" })

// Create a measure.
const measureTen = globalStats.createMeasureInt64('example.com/ten', MeasureUnit.UNIT, "10: always and forever!")

// Create a "last value" view for the measure.
const viewTenLast = globalStats.createView(
    'example.com/ten_last',
    measureTen,
    AggregationType.LAST_VALUE,
    [tagKeyApp],
    'Last value of ten.'
)

globalStats.registerView(viewTenLast)

// Record a value for the measure once per second.
setInterval(() => {
    globalStats.record([{measure: measureTen, value: 10}], tagMapTen);
}, 1000);
``` 

## Useful Links
- To know more about Zenoss, visit: <https://zenoss.com>
- For more information on OpenCensus, visit: <https://opencensus.io/>
- To checkout the OpenCensus for Node.js, visit: <https://github.com/census-instrumentation/opencensus-node>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## v0.0.6
- Fix incorrect setting of zenoss-api-key header.

## v0.0.4
- Change source-type to zenoss/opencensus-node-exporter.
- Small improvements to create-policy.sh script.
- Add LICENSE file to match package's Apache-2.0 license.

## v0.0.3
- Remove leftover debug logging.

## v0.0.2
- Relax @opencensus/core requirement to v0.0.9.

## v0.0.1
- Initial alpha release.


create-policy.sh
#!/usr/bin/env bash
#
# This script provides an example of creating (and deleting) Zenoss policy
# to handle the tagged metrics sent by @zenoss/opencensus-node-exporter.
#
# The policy assumes you have set the "source" option of your exporter, and
# that all of your metrics are tagged with an "app" tag, or you're adding the
# "app" tag via the exporter's "extraTags" option.
#
# This will result in an entity being created that represents the source/app
# combination, and metrics being attributed to that entity.
#
# ZENOSS_ADDRESS can be overridden, but defaults to https://api.zenoss.io
#
# You must set the ZENOSS_API_KEY environment variable.

ZENOSS_ADDRESS="${ZENOSS_ADDRESS:-https://api.zenoss.io}"

if [ -z "$ZENOSS_API_KEY" ]; then
    echo "ZENOSS_API_KEY must be set."
    exit 1
fi

# Create API helper functions.
zenoss-api-post () {
  echo "POST $1"
  curl "$ZENOSS_ADDRESS$1" \
    -H "zenoss-api-key: $ZENOSS_API_KEY" \
    -k -s -X POST -d @-
  echo
}

zenoss-api-put () {
  echo "PUT $1"
  curl "$ZENOSS_ADDRESS$1" \
    -H "zenoss-api-key: $ZENOSS_API_KEY" \
    -k -s -X PUT -d @-
  echo
}

zenoss-api-delete () {
  echo "DELETE $1"
  curl "$ZENOSS_ADDRESS$1" \
    -H "zenoss-api-key: $ZENOSS_API_KEY" \
    -k -s -X DELETE
  echo
}

# Delete policies if first argument is -d.
if [ "$1" == "-d" ]; then
    zenoss-api-delete /v1/policy/custom/datasources/opencensus-node-exporter
    zenoss-api-delete /v1/policy/custom/ingests/metricIngest_opencensus
    zenoss-api-delete /v1/policy/custom/ingests/modelIngest_opencensus
    exit 0
fi

# Create metric ingest policy.
zenoss-api-post /v1/policy/custom/ingests << EOF
{
    "name": "metricIngest_opencensus",
    "requiredKeys": ["source", "source-type", "app"],
    "dimensionKeys": ["source", "app"],
    "metadataKeys": ["source-type"],
    "useAllKeysAsDims": false,
    "excludeInternalFromAllKeysAsDims": true,
    "generateEntityId": true,
    "entityDimensionKeys": ["source", "app"],
    "contributeModelInfo": true
}
EOF

# Create model ingest policy.
zenoss-api-post /v1/policy/custom/ingests << EOF
{
    "name": "modelIngest_opencensus",
    "requiredKeys": ["source", "source-type", "app"],
    "fieldTransforms": [
        {
            "operation": "COPY",
            "sourceKey": "source",
            "targetKey": "name"
        }
    ],
    "dimensionKeys": ["source", "app"],
    "metadataKeys": ["source-type", "name"],
    "useAllKeysAsDims": false,
    "excludeInternalFromAllKeysAsDims": true,
    "generateEntityId": true,
    "entityDimensionKeys": ["source", "app"]
}
EOF

# Create datasource that uses the metric and model ingest policies.
zenoss-api-post /v1/policy/custom/datasources << EOF
{
    "name": "opencensus-node-exporter",
    "sourceType": "zenoss/opencensus-node-exporter",
    "policyReferences": [
        {
            "dataType": 2,
            "policyType": 2,
            "policyName": "metricIngest_opencensus"
        },
        {
            "dataType": 3,
            "policyType": 2,
            "policyName": "modelIngest_opencensus"
        }
    ]
}
EOF


src/zenoss-stats.ts
import {
    DistributionValue,
    ExporterConfig,
    logger,
    Logger,
    Measurement,
    MetricDescriptorType,
    MetricProducerManager,
    Metrics,
    StatsEventListener, SummaryValue,
    TagKey,
    TagValue, Timestamp,
    View,
} from '@opencensus/core';

const fetch = require('node-fetch');

export interface ZenossStatsExporterOptions extends ExporterConfig {
    // address: Zenoss API endpoint address.
    address?: string;

    // apiKey: Zenoss API key.
    apiKey?: string;

    // source: recommended tag identifying the source of data.
    source?: string;

    // extraTags: optional map of extra tags to add to every metric.
    extraTags?: { [key: string]: string };

    // period: how often (in milliseconds) to send metrics to Zenoss.
    period?: number;

    // logger: optional custom Logger to be used by the exporter.
    logger?: Logger;
}

export class ZenossStatsExporter implements StatsEventListener {
    static readonly DEFAULT_ADDRESS: string = "https://api.zenoss.io"
    static readonly DEFAULT_PERIOD: number = 60000;
    static readonly DEFAULT_SOURCE_TYPE: string = "zenoss/opencensus-node-exporter";

    static readonly API_KEY_FIELD: string = "zenoss-api-key";
    static readonly SOURCE_FIELD: string = "source";
    static readonly SOURCE_TYPE_FIELD: string = "source-type";
    static readonly DESCRIPTION_FIELD: string = "description";
    static readonly UNITS_FIELD: string = "units";

    private readonly address: string;
    private readonly apiKey?: string;
    private readonly extraTags: { [key: string]: string };
    private readonly period: number;
    private readonly logger: Logger;

    private timer!: NodeJS.Timer;

    constructor(options: ZenossStatsExporterOptions) {
        this.address = options.address || ZenossStatsExporter.DEFAULT_ADDRESS;
        this.apiKey = options.apiKey;
        this.extraTags = options.extraTags || {};
        this.period = options.period || ZenossStatsExporter.DEFAULT_PERIOD;
        this.logger = options.logger || logger.logger();

        if (!this.extraTags[ZenossStatsExporter.SOURCE_TYPE_FIELD]) {
            this.extraTags[ZenossStatsExporter.SOURCE_TYPE_FIELD] = ZenossStatsExporter.DEFAULT_SOURCE_TYPE;
        }

        if (options.source) {
            if (!this.extraTags[ZenossStatsExporter.SOURCE_FIELD]) {
                this.extraTags[ZenossStatsExporter.SOURCE_FIELD] = options.source
            }
        }
    }

    start(): void {
        this.timer = setInterval(async () => {
            try {
                await this.export();
            } catch (err) {
                this.logger.error("failed to export metrics: %s", err)
            }
        }, this.period);
    }

    async export() {
        const now = Date.now() / 1000;
        const taggedMetrics: TaggedMetric[] = [];
        const metricProducerManager: MetricProducerManager = Metrics.getMetricProducerManager();

        const addMetric = (name: string, timestamp: Timestamp, value: number, tags: { [k: string]: string }) => {
            taggedMetrics.push({
                metric: name,
                timestamp: (timestamp.seconds || now) * 1000,
                value: value,
                tags: tags,
            })
        }

        for (const metricProducer of metricProducerManager.getAllMetricProducer()) {
            for (const metric of metricProducer.getMetrics()) {
                for (const timeSeries of metric.timeseries) {
                    const tags: { [key: string]: string } = {};

                    for (let i = 0; i < timeSeries.labelValues.length; i++) {
                        const labelKey = metric.descriptor.labelKeys[i];
                        const labelValue = timeSeries.labelValues[i].value;
                        if (labelKey && labelValue !== null && labelValue !== undefined) {
                            tags[labelKey.key] = labelValue;
                        }
                    }

                    if (metric.descriptor.description) {
                        tags[ZenossStatsExporter.DESCRIPTION_FIELD] = metric.descriptor.description;
                    }

                    if (metric.descriptor.unit) {
                        tags[ZenossStatsExporter.UNITS_FIELD] = metric.descriptor.unit;
                    }

                    // extraTags override metric tags.
                    for (const k in this.extraTags) {
                        tags[k] = this.extraTags[k];
                    }

                    for (const point of timeSeries.points) {
                        switch (metric.descriptor.type) {
                            case MetricDescriptorType.CUMULATIVE_DOUBLE:
                            case MetricDescriptorType.GAUGE_DOUBLE:
                            case MetricDescriptorType.CUMULATIVE_INT64:
                            case MetricDescriptorType.GAUGE_INT64:
                                addMetric(metric.descriptor.name, point.timestamp, point.value as number, tags);
                                break;

                            case MetricDescriptorType.CUMULATIVE_DISTRIBUTION:
                            case MetricDescriptorType.GAUGE_DISTRIBUTION:
                                const distribution = point.value as DistributionValue
                                const dValues: { [key: string]: number } = {
                                    count: distribution.count,
                                    sum: distribution.sum,
                                    ss: distribution.sumOfSquaredDeviation,
                                    mean: distribution.count === 0 ? 0 : distribution.sum / distribution.count,
                                };

                                for (const k in dValues) {
                                    addMetric(metric.descriptor.name + "/" + k, point.timestamp, dValues[k], tags);
                                }

                                break;

                            case MetricDescriptorType.SUMMARY:
                                const summary = point.value as SummaryValue
                                const sValues: { [key: string]: number } = {
                                    count: summary.count,
                                    sum: summary.sum,
                                }

                                for (const k in sValues) {
                                    addMetric(metric.descriptor.name + "/" + k, point.timestamp, sValues[k], tags)
                                }

                                break

                            default:
                                this.logger.warn("unknown metric type: %s", metric.descriptor.type)
                        }
                    }
                }
            }
        }

        if (taggedMetrics.length > 0) {
            await this.sendTaggedMetrics(taggedMetrics);
        }
    }

    private async sendTaggedMetrics(taggedMetrics: TaggedMetric[]) {
        const metricsUrl = this.address + "/v1/data-receiver/metrics";
        const headers: {[key: string]: string} = {};
        if (this.apiKey) {
            headers[ZenossStatsExporter.API_KEY_FIELD] = this.apiKey;
        }

        try {
            const response = await fetch(metricsUrl, {
                method: 'post',
                headers: headers,
                body: JSON.stringify({taggedMetrics}),
                timeout: this.period,
            });

            if (!response.ok) {
                const text = await response.text();
                this.logger.error(
                    "failed to send metrics: %s %s: %s",
                    response.status,
                    response.statusText,
                    text);
            } else {
                const json = await response.json();
                this.logger.debug("sent metrics:", json)
            }
        } catch (error) {
            this.logger.error("failed to send metrics: %s", error);
        }
    }

    stop(): void {
        clearInterval(this.timer);
    }

    // Deprecated by opencensus-node. Exists only to satisfy StatsEventListener interface.
    onRegisterView(view: View): void {}

    // Deprecated by opencensus-node. Exists only to satisfy StatsEventListener interface.
    onRecord(views: View[], measurement: Measurement, tags: Map<TagKey, TagValue>) {}
}

interface TaggedMetric {
    metric: string;
    timestamp: number;
    value: number;
    tags: { [k: string]: string };
}
